[CLI] Add nested vector arg support, docs (#7709)

alnoki opened the pull request:
@davidiw @gregnazario @movekevin

Addresses #7709 

# Summary of changes

* Implement recursive vector parsing for `ArgWithType`
* Abstract out `ArgWithTypeVec` field in accordance with [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself) design principles
* Add Move example with multiple nested vector permutations
* Add CLI docs for running example Move code

# Vector syntax

Previously, the CLI was restricted to vector arguments with a maximum depth of 1, per the below syntax:

`vector<u8>:2,5,3`

Notably, this schema does *not* support doubly nested vectors of the form `vector<vector<u8>>`, as there is no way to delimit inner vectors.

However, since Move vectors must all use the same inner type, the new schema employs JSON array syntax of the form:

`"u8:[[1, 2], [], [4, 5, 6]]"`

Note here the use of quotes to escape spaces and square braces, as required for typical shell interpreters like `zsh`.

Hence as described in the docs example from this PR, a Move function can be invoked via syntax of the form:

```zsh
aptos move run \
    --function-id <deploy_address>::cli_args::set_vals \
    --private-key-file <keyfile> \
    --args \
        u8:123 \
        "bool:[false, true, false, false]" \
        'address:[["0xace", "0xbee"], ["0xcad"], ["0xdee"], []]'
```

# Additional example

See also testnet transaction [499639189](https://explorer.aptoslabs.com/txn/499639189/payload?network=testnet) for a public entry function invocation with a `vector<vector<u64>>` argument, run from the CLI via:

```zsh
aptos move run \
    --function-id $testnet_address::incentives::update_incentives \
    --private-key-file tesnet_account.key \
    --type-args 0x1::aptos_coin::AptosCoin \
    --args \
        u64:1 \
        u64:1 \
        u64:1 \
        u64:2000 \
        "u64:[[10000, 0, 7], [8333, 1, 6], [7692, 2, 5], [7143, 3, 4], [6667, 4, 3], [6250, 5, 2], [5882, 6, 1]]"
```

# Local checks

The following commands were run in the below directories:

* `aptos-core/`

    ```zsh
    pre-commit run --all-files
    scripts/rust_lint.sh
    ```

* `aptos-core/crates/aptos/`

    ```zsh
    cargo test
    ```

* `aptos-core/developer-docs-site/`

    ```
    pnpm spellcheck
    pnpm lint
    ```

gregnazario commented on crates/aptos/src/move_tool/mod.rs:
```diff
@@ -1418,137 +1404,140 @@ impl FromStr for FunctionArgType {
+    /// Vectors may be specified using JSON array literal syntax (you may need to escape this with
+    /// quotes based on your shell interpreter)
+    ///
+    /// Example: `address:0x1 bool:true u8:0 u256:1234 "bool:[true, false]" 'address:[["0xace", "0xbee"], []]'`
```
This is pretty clean, might have to make a version bump and note that it's a breaking change in the CHANGELOG.md

alnoki replied:
Can do, I'm glad to add these unless you would rather handle it.

gregnazario commented on crates/aptos/src/move_tool/mod.rs:
```diff
@@ -1418,137 +1404,140 @@ impl FromStr for FunctionArgType {
     }
 }
 
+/// Does not support string arguments that contain the following characters:
+///
+/// * `,`
+/// * `[`
+/// * `]`
```
We'll need to add escaping soon

alnoki replied:
Agreed.

gregnazario reviewed (commented):
Pretty nice, I'll have to test it out and all, been very busy.  Noting that this is a breaking change.  @0xjinn can take a look as well

alnoki commented:
@gregnazario I've addressed your comments above and resolved merge conflicts, so this should be ready to land!

Let me know if you or @0xjinn have anything to discuss

gregnazario commented on developer-docs-site/docs/tools/aptos-cli-tool/use-aptos-cli.md:
```diff
@@ -719,6 +719,54 @@ The `peer_config.yaml` file will be created in your current working directory, w
 
 Move examples can be found in the [Move section](../../move/move-on-aptos/cli).
 
+You can also pass multiply-nested vector arguments, like in the `cli_args` example from [move-examples](https://github.com/aptos-labs/aptos-core/tree/main/aptos-move/move-examples):
```
`multi-nested` vs `multiply-nested` sounds like you're multiplying :)

alnoki replied:
Resolved in 329d1e5

gregnazario commented on crates/aptos/src/move_tool/mod.rs:
```diff
@@ -1223,130 +1214,134 @@ impl Display for FunctionArgType {
-    let args = args.split(',');
-    for arg in args {
-        parsed_args.push(parse(arg)?);
+// TODO use from move_binary_format::file_format_common if it is made public.
+fn write_u64_as_uleb128(binary: &mut Vec<u8>, mut val: usize) {
+    loop {
+        let cur = val & 0x7F;
+        if cur != val {
+            binary.push((cur | 0x80) as u8);
+            val >>= 7;
+        } else {
+            binary.push(cur as u8);
+            break;
+        }
     }
```
Is there a reason we're rewriting the BCS encoding logic?

Should be able to just put it in a vector and then BCS encode it.

alnoki replied:
> Should be able to just put it in a vector and then BCS encode it.

```rust
            serde_json::Value::Array(_) => {
                // Vector of sub-argument BCS representations.
                let mut sub_args: Vec<Vec<u8>> = vec![]; // <------------------------------------- 1
                let mut common_sub_arg_depth = None;
                // Loop over all of the vector's sub-arguments, which may also be vectors:
                for sub_arg in arg.as_array().unwrap() {
                    let ArgWithType {
                        _ty: _,
                        _vector_depth: sub_arg_depth,
                        arg: sub_arg_bcs,
                    } = self.parse_arg_json(sub_arg)?;
                    // Verify all sub-arguments have same depth.
                    if let Some(check_depth) = common_sub_arg_depth {
                        if check_depth != sub_arg_depth {
                            return Err(CliError::CommandArgumentError(
                                "Variable vector depth".to_string(),
                            ));
                        }
                    };
                    common_sub_arg_depth = Some(sub_arg_depth);
                    // Push back sub-argument BCS.
                    sub_args.push(sub_arg_bcs); // <---------------------------------------------- 2
                }
                // Default sub-argument depth is 0 for when no sub-arguments were looped over.
                Ok(ArgWithType {
                    _ty: self.clone(),
                    _vector_depth: common_sub_arg_depth.unwrap_or(0) + 1,
                    arg: bcs::to_bytes(&sub_args).map_err(|err| CliError::BCS("arg", err))?,
                    // ^ ------------------------------------------------------------------------- 3
                })
            },
            serde_json::Value::Null => {
                Err(CliError::CommandArgumentError("Null argument".to_string()))
            },
            serde_json::Value::Object(_) => Err(CliError::CommandArgumentError(
                "JSON object argument".to_string(),
            )),
```

At first glance it might appear that the above algorithm does what you are asking:

1. Declare a vector of sub-argument BCS encodings (`Vec<Vec<u8>>`)
2. Push all sub-argument BCS encodings onto the vector
3. Convert the entire `Vec<Vec<u8>>` to BCS

However, the problem with this approach is that the BCS encoding logic generates not only a `uleb128` for the top-level vector, but also for each nested vector within. This shouldn't happen, however, because the relevant BCS `Vec<u8>` encoding of each sub-argument has already been calculated (including with the leading `uleb128`) during recursion. The problem here is that when the BCS  encoder parses `Vec<Vec<u8>>` it doesn't treat each subvector `Vec<u8>` as already being in BCS form.

The reason I've rewritten the BCS logic is because it is not publicly available. Note that the same thing is already happening in [`transaction_arg_validation.rs`](https://github.com/aptos-labs/aptos-core/blob/6aed685d6c5288fa039a9ce59908fd7809d861c1/aptos-move/aptos-vm/src/verifier/transaction_arg_validation.rs#L374)

gregnazario commented on crates/aptos/src/move_tool/mod.rs:
```diff
@@ -1223,130 +1214,134 @@ impl Display for FunctionArgType {
+                arg: self.parse_arg_str(value.as_str())?,
+            }),
+            serde_json::Value::Array(_) => {
+                let mut bcs: Vec<u8> = vec![]; // BCS representation of argument.
+                let mut common_sub_arg_depth = None;
+                // Prepend argument sequence length to BCS bytes vector.
+                write_u64_as_uleb128(&mut bcs, arg.as_array().unwrap().len());
+                // Loop over all of the vector's sub-arguments, which may also be vectors:
+                for sub_arg in arg.as_array().unwrap() {
+                    let ArgWithType {
+                        _ty: _,
+                        _vector_depth: sub_arg_depth,
+                        arg: mut sub_arg_bcs,
+                    } = self.parse_arg_json(sub_arg)?;
+                    // Verify all sub-arguments have same depth.
+                    if let Some(check_depth) = common_sub_arg_depth {
+                        if check_depth != sub_arg_depth {
+                            return Err(CliError::CommandArgumentError(
+                                "Variable vector depth".to_string(),
+                            ));
+                        }
+                    };
+                    common_sub_arg_depth = Some(sub_arg_depth);
+                    bcs.append(&mut sub_arg_bcs); // Append sub-argument BCS.
```
My only concern is that we're reimplementing BCS, but probably points out that maybe BCS needs a more interactive streaming approach in Rust like you're doing here.

alnoki replied:
Agreed. I go into more depth in the above comment https://github.com/aptos-labs/aptos-core/pull/7790#discussion_r1178402832

gregnazario commented on crates/aptos/src/test/mod.rs:
```diff
@@ -591,13 +593,15 @@ impl CliTestFramework {
-                    ArgWithType::u64(commission_percentage),
-                    ArgWithType::bytes(vec![]),
-                ],
+                arg_vec: ArgWithTypeVec {
+                    args: vec![
+                        ArgWithType::address(self.account_id(operator_index)),
+                        ArgWithType::address(self.account_id(voter_index)),
+                        ArgWithType::u64(amount),
+                        ArgWithType::u64(commission_percentage),
+                        ArgWithType::bytes(vec![]),
+                    ],
```
We should probably add a unit test for this function now (I feel like I've avoided it too long), that checks the multi-nested BCS, and ensures it can be deserialized as the type expected.

alnoki replied:
Agreed. I am also working on a PR that will extend this functionality to include JSON input files, and it looks like it'll involve some minor refactors to struct nesting. Once we can get that in perhaps it'll be appropriate to add more unit testing?

gregnazario reviewed (approved):
I'm going to approve this @0xjinn can you take another look at this.

0xjinn commented on crates/aptos/src/move_tool/mod.rs:
```diff
@@ -1223,130 +1214,134 @@ impl Display for FunctionArgType {
+    }
+
+    /// Recursively parse argument JSON into BCS representation.
+    pub fn parse_arg_json(&self, arg: &serde_json::Value) -> CliTypedResult<ArgWithType> {
```
nit: Would be great if we can unit test and safeguard the parser

alnoki replied:
Agreed. As I mention in https://github.com/aptos-labs/aptos-core/pull/7790#discussion_r1178408642 I'm working on a separate PR that will extend support to JSON input files as well. Perhaps once both input types are supported an end-to-end test will be in order?

alnoki commented:
@gregnazario I believe I've addressed all of your comments/suggestions.

@0xjinn I believe I've addressed your proposition.

Is there anything else needed to get an approving review from @0xjinn and land in `main`?

0xjinn commented:
I think this looks good on my end. I did manual tested it and ensured the [example transaction](https://explorer.aptoslabs.com/txn/0xef51aff503f9c81f0d033c7c253ce50792a6f996a8018b26494e7e0a9a47d51f/userTxnOverview?network=testnet) works.

I can follow up with PR to update the changelog.

cc: @gregnazario for final look

davidiw commented:
claims it needs to be rebased

alnoki commented:
> claims it needs to be rebased

@davidiw I just merged in `main` without conflicts, perhaps the notice was due to the time since branching?

Either way @gregnazario this should be ready to land

alnoki commented:
@davidiw @gregnazario I've further squashed the PR so it reflects as a single commit against `main`

There still are no merge conflicts, so I'm not sure if there's anything else I can do to massage the diff history

lightmark commented:
love this @alnoki

alnoki commented:
> love this @alnoki

@lightmark see also https://github.com/aptos-labs/aptos-core/pull/8054
